core/src/main/scala/kafka/message/Message.scala

```diff
@@ -36,83 +37,139 @@ object Message {
- * 4. 4 byte key length, containing length K
- * 5. K byte key
- * 6. 4 byte payload length, containing length V
- * 7. V byte payload
+ * 4. (Optional) 8 byte timestamp only if "magic" identifier is greater than 0
+ * 5. 4 byte key length, containing length K
+ * 6. K byte key
+ * 7. 4 byte payload length, containing length V
+ * 8. V byte payload
  *
  * Default constructor wraps an existing ByteBuffer with the Message object with no change to the contents.
  */
-class Message(val buffer: ByteBuffer) {
+class Message(val buffer: ByteBuffer,
+              private val wrapperMessageTimestamp: Option[Long] = None,
+              private val wrapperMessageTimestampType: Option[TimestampType] = None) {
   
   import kafka.message.Message._
-  
+
   /**
    * A constructor to create a Message
    * @param bytes The payload of the message
    * @param codec The compression codec used on the contents of the message (if any)
+   * @param timestamp The timestamp of the message.
+   * @param timestampType The timestamp type of the message. Default to CreateTime
    * @param key The key of the message (null, if none)
    * @param payloadOffset The offset into the payload array used to extract payload
    * @param payloadSize The size of the payload to use
    */
   def this(bytes: Array[Byte], 
-           key: Array[Byte],            
+           key: Array[Byte],
+           timestamp: Long,
+           timestampType: TimestampType = CreateTime,
            codec: CompressionCodec, 
            payloadOffset: Int, 
-           payloadSize: Int) = {
-    this(ByteBuffer.allocate(Message.CrcLength + 
-                             Message.MagicLength + 
-                             Message.AttributesLength + 
+           payloadSize: Int,
+           magicValue: Byte) = {
+    this(ByteBuffer.allocate(Message.CrcLength +
+                             Message.MagicLength +
+                             Message.AttributesLength +
+                             (if (magicValue == Message.MagicValue_V0) 0
+                              else Message.TimestampLength) +
                              Message.KeySizeLength + 
                              (if(key == null) 0 else key.length) + 
                              Message.ValueSizeLength + 
                              (if(bytes == null) 0 
                               else if(payloadSize >= 0) payloadSize 
                               else bytes.length - payloadOffset)))
+    validateTimestampAndMagicValue(timestamp, magicValue)
     // skip crc, we will fill that in at the end
     buffer.position(MagicOffset)
-    buffer.put(CurrentMagicValue)
+    buffer.put(magicValue)
     var attributes: Byte = 0
     if (codec.codec > 0)
       attributes =  (attributes | (CompressionCodeMask & codec.codec)).toByte
     buffer.put(attributes)
+    // Only put timestamp when "magic" value is greater than 0
```

Comment:
Hi Jun, I actually hesitated a little here. It seems this constructor should only be used by producer, so the timestamp should always be CreateTime. I added the timestamp type to constructor pretty lately because  ChecksumMessageFormatter also needs to construct message in order to compute checksum. 

Currently ChecksumMessageFormatter only takes key and value and always assume the compression type to be NoCompression. This works because all the messages, including inner messages of compressed messages, should not have compression codec.

However, because timestamp type can be different from message to message, we need to include timestamp type when computing checksum. We also need compression type because for compressed messages, inner message timestamp type is always CreateTime even when the timestamp type of the message is LogAppendTime defined by wrapper message.

I did not find any usage of ChecksumMessageFormatter. I asked Joel about this class, and it looks we used to use it in system test, but now we are not using it anymore. Do you think we can simply remove this class?